### PR TITLE
Service get best answer

### PR DESCRIPTION
**Pull request by GeoffreyHuck**
Fixes 874

### CODE REVIEW

**Comment by GeoffreyHuck** [comment]
Fixes #874

**Review by smadbe — CHANGES_REQUESTED** [review]
Looks good overall 👍 

For each remark, answer "fixed" when done (or explain if you don't fix) so that I can re-review and I will mark them "resolved" myself on GH.

File: app/api/answers/get_best.feature [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    | 1          | 11             | 210     |
+  And the database has the following table 'answers':
+    | id  | author_id | participant_id | attempt_id | item_id | type       | state   | answer   | created_at          |
+    | 101 | 11        | 11             | 1          | 200     | Submission | Current | print(1) | 2017-05-29 06:38:38 |
+    | 102 | 11        | 11             | 2          | 200     | Submission | Current | print(2) | 2017-05-29 07:38:38 |
+    | 103 | 11        | 11             | 3          | 200     | Submission | Current | print(3) | 2017-05-29 08:38:38 |
+    | 104 | 11        | 13             | 1          | 210     | Submission | Current | print(4) | 2017-05-29 09:38:38 |
+    | 105 | 11        | 13             | 1          | 210     | Submission | Current | print(5) | 2017-05-29 10:38:38 |
+    | 106 | 11        | 13             | 1          | 210     | Submission | Current | print(6) | 2017-05-29 11:38:38 |
+    | 107 | 11        | 11             | 1          | 210     | Submission | Current | print(7) | 2017-05-29 08:38:38 |
```
[the remark this answers is not in the document]
Comment (GeoffreyHuck) [reply app/api/answers/get_best.feature]: fixed for all tests

File: app/api/answers/get_best.feature [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    | 23       | 210     | content_with_descendants | answer                |
+  And the database has the following table 'results':
+    | attempt_id | participant_id | item_id |
+    | 1          | 11             | 200     |
+    | 2          | 11             | 200     |
+    | 3          | 11             | 200     |
+    | 1          | 13             | 210     |
+    | 1          | 11             | 210     |
+  And the database has the following table 'answers':
+    | id  | author_id | participant_id | attempt_id | item_id | type       | state   | answer   | created_at          |
```
[the remark this answers is not in the document]
Comment (GeoffreyHuck) [reply app/api/answers/get_best.feature]: fixed

File: app/api/answers/get_best.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	if apiError != service.NoError {
+		return apiError
+	}
+
+	user := srv.GetUser(httpReq)
+	store := srv.GetStore(httpReq)
+	var result []map[string]interface{}
+
+	var bestAnswerQuery *database.DB
+	if watchedGroupID != 0 && watchedGroupIDSet {
```
[the remark this answers is not in the document]
Comment (GeoffreyHuck) [reply app/api/answers/get_best.go]: fixed, it was useless. Wouldn't it be even better to rename watchedGroupIDSet into "ok" to follow the "comma ok idiom" ?

File: app/api/answers/get_best.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		// check 'can_watch'>='answer' permission on the answers.item_id
+		itemPerms := store.Permissions().MatchingUserAncestors(user).
+			WherePermissionIsAtLeast("watch", "answer").
+			Where("permissions.item_id = answers.item_id").
+			Select("1").
+			Limit(1)
+
+		// check if able to watch the participant
+		participantPerms := store.ActiveGroupAncestors().ManagedByUser(user).
+			Joins("JOIN `groups` ON groups.id = groups_ancestors_active.child_group_id").
```
[the remark this answers is not in the document]
Comment (GeoffreyHuck) [reply app/api/answers/get_best.go]: fixed, it came from an other service. I deleted it there too.

File: app/api/answers/get_best.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+			Limit(1)
+
+		bestAnswerQuery = store.Answers().
+			Where(`?`, itemPerms.SubQuery()).
+			Where(`?`, participantPerms.SubQuery()).
+			Where("participant_id = ?", watchedGroupID)
+	} else {
+		bestAnswerQuery = store.Answers().
+			Visible(user).
+			Where("author_id = ?", user.GroupID)
```
[the remark this answers is not in the document]
Comment (GeoffreyHuck) [reply app/api/answers/get_best.go]: fixed. I also added a test that covers the case.

I removed `.Visible(user)` that makes more checks than required and replaced it with `WherePermissionIsAtLeast("view", "content")` instead.

File: app/api/answers/get_best.feature [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    | 107       | 98    | 2019-05-29 06:38:38 |
+
+  Scenario: User has access to the item and retrieves his best answer
+    Given I am the user with id "11"
+    When I send a GET request to "/items/200/best-answer"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    {
+      "id": "103",
```
[the remark this answers is not in the document]
Comment (GeoffreyHuck) [reply app/api/answers/get_best.feature:79]: fixed

File: app/api/answers/get_best.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (smadbe) [reply app/api/answers/get_best.go]: Yes, don't know why it has not been done initially.

**Review by smadbe — APPROVED** [review]
great 👍

**Pull request merged**